src/Components/Endpoints/src/Builder/ResourcePreloadCollection.cs

```diff
@@ -0,0 +1,94 @@
+        {
+            var headers = new List<(string? Group, int Order, string Value)>();
+            foreach (var asset in assets)
+            {
+                if (asset.Properties == null)
+                {
+                    continue;
+                }
+
+                // Use preloadgroup=webassembly to identify assets that should to be preloaded
+                string? group = null;
+                foreach (var property in asset.Properties)
+                {
+                    if (property.Name.Equals("preloadgroup", StringComparison.OrdinalIgnoreCase))
+                    {
+                        group = property.Value ?? string.Empty;
+                        break;
+                    }
+                }
+
+                if (group == null)
+                {
+                    continue;
+                }
+
+                var header = new StringBuilder();
+                header.Append('<');
+                header.Append(asset.Url);
+                header.Append('>');
+
+                int order = 0;
+                foreach (var property in asset.Properties)
+                {
+                    if (property.Name.Equals("preloadrel", StringComparison.OrdinalIgnoreCase))
+                    {
+                        header.Append("; rel=").Append(property.Value);
+                    }
+                    else if (property.Name.Equals("preloadas", StringComparison.OrdinalIgnoreCase))
+                    {
+                        header.Append("; as=").Append(property.Value);
+                    }
+                    else if (property.Name.Equals("preloadpriority", StringComparison.OrdinalIgnoreCase))
+                    {
+                        header.Append("; fetchpriority=").Append(property.Value);
+                    }
+                    else if (property.Name.Equals("preloadcrossorigin", StringComparison.OrdinalIgnoreCase))
+                    {
+                        header.Append("; crossorigin=").Append(property.Value);
+                    }
+                    else if (property.Name.Equals("integrity", StringComparison.OrdinalIgnoreCase))
+                    {
+                        header.Append("; integrity=\"").Append(property.Value).Append('"');
+                    }
+                    else if (property.Name.Equals("preloadorder", StringComparison.OrdinalIgnoreCase))
+                    {
+                        if (!int.TryParse(property.Value, out order))
+                        {
+                            order = 0;
+                        }
+                    }
+                }
+
+                if (header != null)
+                {
+                    headers.Add((group, order, header.ToString()));
+                }
+            }
+
+            foreach (var group in headers.GroupBy(h => h.Group))
+            {
+                _storage[group.Key ?? string.Empty] = group.OrderBy(h => h.Order).Select(h => h.Value).ToArray();
```

Comment:
Should the order here "override" the fetch priority?

Also, rather than starting with a list, could we start already with a Dictionary<string, List<string>> or something like that?

Reply:
The idea is that within the same fetch priority, you can define the order, as throttling may apply